File: TarsEngine.FSharp.Core/Analysis/SecurityIssueDetector.fs
namespace TarsEngine.FSharp.Core.Analysis

open System
open System.Text.RegularExpressions
open Microsoft.Extensions.Logging

/// <summary>
/// Implementation of ISecurityIssueDetector for C# and F# languages.
/// </summary>
type SecurityIssueDetector(logger: ILogger<SecurityIssueDetector>) =
    
    /// <summary>
    /// Gets the language supported by this detector.
    /// </summary>
    member _.Language = "csharp"
    
    /// <summary>
    /// Detects issues in the provided content.
    /// </summary>
    /// <param name="content">The source code content.</param>
    /// <returns>A list of detected issues.</returns>
    member _.DetectIssues(content: string) =
        try
            logger.LogInformation("Detecting security issues")
            
            // Detect SQL injection vulnerabilities
            let sqlInjectionIssues = this.DetectSqlInjection(content)
            
            // Detect cross-site scripting vulnerabilities
            let xssIssues = this.DetectXss(content)
            
            // Detect insecure cryptography
            let insecureCryptographyIssues = this.DetectInsecureCryptography(content)
            
            // Detect hardcoded credentials
            let hardcodedCredentialsIssues = this.DetectHardcodedCredentials(content)
            
            // Combine all issues
            List.concat [
                sqlInjectionIssues
                xssIssues
                insecureCryptographyIssues
                hardcodedCredentialsIssues
            ]
        with
        | ex ->
            logger.LogError(ex, "Error detecting security issues")
            []
    
    /// <summary>
    /// Detects issues in a file.
    /// </summary>
    /// <param name="filePath">The path to the file.</param>
    /// <returns>A list of detected issues.</returns>
    member this.DetectIssuesInFile(filePath: string) =
        try
            logger.LogInformation("Detecting security issues in file: {FilePath}", filePath)
            
            // Read the file content
            let content = System.IO.File.ReadAllText(filePath)
            
            // Detect issues in the content
            let issues = this.DetectIssues(content)
            
            // Update issues with file path
            issues |> List.map (fun issue -> { issue with FilePath = Some filePath })
        with
        | ex ->
            logger.LogError(ex, "Error detecting security issues in file: {FilePath}", filePath)
            []
    
    /// <summary>
    /// Gets the supported issue types.
    /// </summary>
    /// <returns>A list of supported issue types.</returns>
    member _.GetSupportedIssueTypes() =
        [
            CodeIssueType.Security
        ]
    
    /// <summary>
    /// Detects SQL injection vulnerabilities.
    /// </summary>
    /// <param name="content">The source code content.</param>
    /// <returns>A list of detected issues.</returns>
    member _.DetectSqlInjection(content: string) =
        try
            // Define patterns for SQL injection vulnerabilities
            let patterns = [
                @"string\s+sql\s*=\s*""[^""]*\+\s*[^""]*""", "String concatenation in SQL query"
                @"ExecuteQuery\s*\(\s*""[^""]*\+\s*[^""]*""\s*\)", "String concatenation in ExecuteQuery"
                @"ExecuteNonQuery\s*\(\s*""[^""]*\+\s*[^""]*""\s*\)", "String concatenation in ExecuteNonQuery"
                @"ExecuteScalar\s*\(\s*""[^""]*\+\s*[^""]*""\s*\)", "String concatenation in ExecuteScalar"
                @"ExecuteReader\s*\(\s*""[^""]*\+\s*[^""]*""\s*\)", "String concatenation in ExecuteReader"
                @"SqlCommand\s*\(\s*""[^""]*\+\s*[^""]*""\s*\)", "String concatenation in SqlCommand"
                @"OleDbCommand\s*\(\s*""[^""]*\+\s*[^""]*""\s*\)", "String concatenation in OleDbCommand"
                @"OdbcCommand\s*\(\s*""[^""]*\+\s*[^""]*""\s*\)", "String concatenation in OdbcCommand"
                @"MySqlCommand\s*\(\s*""[^""]*\+\s*[^""]*""\s*\)", "String concatenation in MySqlCommand"
                @"NpgsqlCommand\s*\(\s*""[^""]*\+\s*[^""]*""\s*\)", "String concatenation in NpgsqlCommand"
            ]
            
            // Detect issues using patterns
            patterns
            |> List.collect (fun (pattern, message) ->
                Regex.Matches(content, pattern, RegexOptions.IgnoreCase ||| RegexOptions.Multiline)
                |> Seq.cast<Match>
                |> Seq.map (fun m ->
                    let lineNumber = content.Substring(0, m.Index).Split('\n').Length
                    let columnNumber = m.Index - content.Substring(0, m.Index).LastIndexOf('\n') - 1
                    let codeSnippet = m.Value
                    
                    {
                        IssueType = CodeIssueType.Security
                        Severity = IssueSeverity.Critical
                        Message = $"Potential SQL injection vulnerability: {message}"
                        LineNumber = Some lineNumber
                        ColumnNumber = Some columnNumber
                        FilePath = None
                        CodeSnippet = Some codeSnippet
                        SuggestedFix = Some "Use parameterized queries instead of string concatenation"
                        AdditionalInfo = Map.empty
                    }
                )
                |> Seq.toList
            )
        with
        | ex ->
            logger.LogError(ex, "Error detecting SQL injection vulnerabilities")
            []
    
    /// <summary>
    /// Detects cross-site scripting vulnerabilities.
    /// </summary>
    /// <param name="content">The source code content.</param>
    /// <returns>A list of detected issues.</returns>
    member _.DetectXss(content: string) =
        try
            // Define patterns for XSS vulnerabilities
            let patterns = [
                @"Response\.Write\s*\(\s*[^""]*\s*\)", "Unencoded output in Response.Write"
                @"<%=\s*[^""]*\s*%>", "Unencoded output in <%= %>"
                @"@Html\.Raw\s*\(\s*[^""]*\s*\)", "Unencoded output in Html.Raw"
                @"innerHTML\s*=\s*[^""]*", "Unencoded output in innerHTML"
                @"document\.write\s*\(\s*[^""]*\s*\)", "Unencoded output in document.write"
            ]
            
            // Detect issues using patterns
            patterns
            |> List.collect (fun (pattern, message) ->
                Regex.Matches(content, pattern, RegexOptions.IgnoreCase ||| RegexOptions.Multiline)
                |> Seq.cast<Match>
                |> Seq.map (fun m ->
                    let lineNumber = content.Substring(0, m.Index).Split('\n').Length
                    let columnNumber = m.Index - content.Substring(0, m.Index).LastIndexOf('\n') - 1
                    let codeSnippet = m.Value
                    
                    {
                        IssueType = CodeIssueType.Security
                        Severity = IssueSeverity.Critical
                        Message = $"Potential cross-site scripting vulnerability: {message}"
                        LineNumber = Some lineNumber
                        ColumnNumber = Some columnNumber
                        FilePath = None
                        CodeSnippet = Some codeSnippet
                        SuggestedFix = Some "Use HTML encoding before outputting user input"
                        AdditionalInfo = Map.empty
                    }
                )
                |> Seq.toList
            )
        with
        | ex ->
            logger.LogError(ex, "Error detecting cross-site scripting vulnerabilities")
            []
    
    /// <summary>
    /// Detects insecure cryptography.
    /// </summary>
    /// <param name="content">The source code content.</param>
    /// <returns>A list of detected issues.</returns>
    member _.DetectInsecureCryptography(content: string) =
        try
            // Define patterns for insecure cryptography
            let patterns = [
                @"MD5", "MD5 is a weak hashing algorithm"
                @"SHA1", "SHA1 is a weak hashing algorithm"
                @"DES", "DES is a weak encryption algorithm"
                @"RC2", "RC2 is a weak encryption algorithm"
                @"RC4", "RC4 is a weak encryption algorithm"
                @"ECB", "ECB mode is insecure for encryption"
                @"new RNGCryptoServiceProvider\(\)", "RNGCryptoServiceProvider is deprecated"
                @"Random\(\)", "System.Random is not cryptographically secure"
            ]
            
            // Detect issues using patterns
            patterns
            |> List.collect (fun (pattern, message) ->
                Regex.Matches(content, pattern, RegexOptions.IgnoreCase ||| RegexOptions.Multiline)
                |> Seq.cast<Match>
                |> Seq.map (fun m ->
                    let lineNumber = content.Substring(0, m.Index).Split('\n').Length
                    let columnNumber = m.Index - content.Substring(0, m.Index).LastIndexOf('\n') - 1
                    let codeSnippet = m.Value
                    
                    {
                        IssueType = CodeIssueType.Security
                        Severity = IssueSeverity.Error
                        Message = $"Insecure cryptography: {message}"
                        LineNumber = Some lineNumber
                        ColumnNumber = Some columnNumber
                        FilePath = None
                        CodeSnippet = Some codeSnippet
                        SuggestedFix = Some "Use modern cryptographic algorithms and practices"
                        AdditionalInfo = Map.empty
                    }
                )
                |> Seq.toList
            )
        with
        | ex ->
            logger.LogError(ex, "Error detecting insecure cryptography")
            []
    
    /// <summary>
    /// Detects hardcoded credentials.
    /// </summary>
    /// <param name="content">The source code content.</param>
    /// <returns>A list of detected issues.</returns>
    member _.DetectHardcodedCredentials(content: string) =
        try
            // Define patterns for hardcoded credentials
            let patterns = [
                @"password\s*=\s*""[^""]+""", "Hardcoded password"
                @"pwd\s*=\s*""[^""]+""", "Hardcoded password"
                @"apikey\s*=\s*""[^""]+""", "Hardcoded API key"
                @"api_key\s*=\s*""[^""]+""", "Hardcoded API key"
                @"secret\s*=\s*""[^""]+""", "Hardcoded secret"
                @"connectionString\s*=\s*""[^""]*password=[^""]+""", "Hardcoded password in connection string"
                @"connectionString\s*=\s*""[^""]*pwd=[^""]+""", "Hardcoded password in connection string"
                @"connectionString\s*=\s*""[^""]*user id=[^""]+""", "Hardcoded user ID in connection string"
                @"connectionString\s*=\s*""[^""]*uid=[^""]+""", "Hardcoded user ID in connection string"
                @"token\s*=\s*""[^""]+""", "Hardcoded token"
            ]
            
            // Detect issues using patterns
            patterns
            |> List.collect (fun (pattern, message) ->
                Regex.Matches(content, pattern, RegexOptions.IgnoreCase ||| RegexOptions.Multiline)
                |> Seq.cast<Match>
                |> Seq.map (fun m ->
                    let lineNumber = content.Substring(0, m.Index).Split('\n').Length
                    let columnNumber = m.Index - content.Substring(0, m.Index).LastIndexOf('\n') - 1
                    let codeSnippet = m.Value
                    
                    {
                        IssueType = CodeIssueType.Security
                        Severity = IssueSeverity.Critical
                        Message = $"Hardcoded credentials: {message}"
                        LineNumber = Some lineNumber
                        ColumnNumber = Some columnNumber
                        FilePath = None
                        CodeSnippet = Some codeSnippet
                        SuggestedFix = Some "Store credentials in a secure configuration system"
                        AdditionalInfo = Map.empty
                    }
                )
                |> Seq.toList
            )
        with
        | ex ->
            logger.LogError(ex, "Error detecting hardcoded credentials")
            []
    
    interface IIssueDetector with
        member this.Language = this.Language
        member this.DetectIssues(content) = this.DetectIssues(content)
        member this.DetectIssuesInFile(filePath) = this.DetectIssuesInFile(filePath)
        member this.GetSupportedIssueTypes() = this.GetSupportedIssueTypes()
    
    interface ISecurityIssueDetector with
        member this.DetectSqlInjection(content) = this.DetectSqlInjection(content)
        member this.DetectXss(content) = this.DetectXss(content)
        member this.DetectInsecureCryptography(content) = this.DetectInsecureCryptography(content)
        member this.DetectHardcodedCredentials(content) = this.DetectHardcodedCredentials(content)

File: archive/v1/parked_csharp/Legacy_CSharp_Projects/TarsEngine/Services/SecurityIssueDetector.cs
using System.Text.RegularExpressions;
using Microsoft.Extensions.Logging;
using TarsEngine.Models;
using TarsEngine.Services.Interfaces;

namespace TarsEngine.Services;

/// <summary>
/// Detects security issues in code
/// </summary>
public class SecurityIssueDetector(ILogger<SecurityIssueDetector> logger) : ISecurityIssueDetector
{
    private readonly ILogger<SecurityIssueDetector> _logger = logger;

    /// <inheritdoc/>
    public string Language => "csharp";

    /// <inheritdoc/>
    public CodeIssueType IssueType => CodeIssueType.Security;

    /// <inheritdoc/>
    public List<CodeIssue> DetectIssues(string content, List<CodeStructure> structures)
    {
        var issues = new List<CodeIssue>();

        try
        {
            if (string.IsNullOrWhiteSpace(content))
            {
                return issues;
            }

            issues.AddRange(DetectSqlInjectionVulnerabilities(content));
            issues.AddRange(DetectXssVulnerabilities(content));
            issues.AddRange(DetectHardcodedCredentials(content));
            issues.AddRange(DetectInsecureCryptography(content));
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error detecting security issues in code");
        }

        return issues;
    }

    /// <inheritdoc/>
    public List<CodeIssue> DetectSqlInjectionVulnerabilities(string content)
    {
        var issues = new List<CodeIssue>();

        try
        {
            if (string.IsNullOrWhiteSpace(content))
            {
                return issues;
            }

            // Detect SQL injection vulnerabilities
            var sqlInjectionRegex = new Regex(@"SqlCommand\s*\(\s*[""'].*?\+\s*[^""']+\s*\+", RegexOptions.Compiled);
            var sqlInjectionMatches = sqlInjectionRegex.Matches(content);
            foreach (Match match in sqlInjectionMatches)
            {
                var lineNumber = GetLineNumber(content, match.Index);
                issues.Add(new CodeIssue
                {
                    Type = CodeIssueType.Security,
                    Severity = TarsEngine.Models.IssueSeverity.Critical,
                    Description = "Use parameterized queries or an ORM instead of string concatenation",
                    Location = new CodeLocation
                    {
                        StartLine = lineNumber,
                        EndLine = lineNumber
                    },
                    Title = "SQL Injection Vulnerability"
                });
            }

            // Detect string concatenation in SQL queries
            var stringConcatRegex = new Regex(@"string\s+sql\s*=\s*[""'].*?\+\s*[^""']+\s*\+", RegexOptions.Compiled);
            var stringConcatMatches = stringConcatRegex.Matches(content);
            foreach (Match match in stringConcatMatches)
            {
                var lineNumber = GetLineNumber(content, match.Index);
                issues.Add(new CodeIssue
                {
                    Type = CodeIssueType.Security,
                    Severity = TarsEngine.Models.IssueSeverity.Major,
                    Description = "Use parameterized queries or an ORM instead of string concatenation",
                    Location = new CodeLocation
                    {
                        StartLine = lineNumber,
                        EndLine = lineNumber
                    },
                    Title = "SQL Injection Risk"
                });
            }
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error detecting SQL injection vulnerabilities");
        }

        return issues;
    }

    /// <inheritdoc/>
    public List<CodeIssue> DetectXssVulnerabilities(string content)
    {
        var issues = new List<CodeIssue>();

        try
        {
            if (string.IsNullOrWhiteSpace(content))
            {
                return issues;
            }

            // Detect XSS vulnerabilities
            var xssRegex = new Regex(@"Response\.Write\s*\(\s*[^""']*\s*\)", RegexOptions.Compiled);
            var xssMatches = xssRegex.Matches(content);
            foreach (Match match in xssMatches)
            {
                var lineNumber = GetLineNumber(content, match.Index);
                issues.Add(new CodeIssue
                {
                    Type = CodeIssueType.Security,
                    Severity = TarsEngine.Models.IssueSeverity.Critical,
                    Description = "Use HTML encoding or a templating engine that automatically escapes output",
                    Location = new CodeLocation
                    {
                        StartLine = lineNumber,
                        EndLine = lineNumber
                    },
                    Title = "XSS Vulnerability"
                });
            }

            // Detect unencoded output
            var unencodedRegex = new Regex(@"@Html\.Raw\s*\(\s*[^""']*\s*\)", RegexOptions.Compiled);
            var unencodedMatches = unencodedRegex.Matches(content);
            foreach (Match match in unencodedMatches)
            {
                var lineNumber = GetLineNumber(content, match.Index);
                issues.Add(new CodeIssue
                {
                    Type = CodeIssueType.Security,
                    Severity = TarsEngine.Models.IssueSeverity.Major,
                    Description = "Use @Html.Encode() or @Html.DisplayFor() instead of @Html.Raw()",
                    Location = new CodeLocation
                    {
                        StartLine = lineNumber,
                        EndLine = lineNumber
                    },
                    Title = "Unencoded Output"
                });
            }
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error detecting XSS vulnerabilities");
        }

        return issues;
    }

    /// <inheritdoc/>
    public List<CodeIssue> DetectHardcodedCredentials(string content)
    {
        var issues = new List<CodeIssue>();

        try
        {
            if (string.IsNullOrWhiteSpace(content))
            {
                return issues;
            }

            // Detect hardcoded credentials
            var credentialsRegex = new Regex(@"(password|pwd|passwd|secret|key|token|apikey)\s*=\s*[""'][^""']+[""']", RegexOptions.IgnoreCase | RegexOptions.Compiled);
            var credentialsMatches = credentialsRegex.Matches(content);
            foreach (Match match in credentialsMatches)
            {
                var lineNumber = GetLineNumber(content, match.Index);
                issues.Add(new CodeIssue
                {
                    Type = CodeIssueType.Security,
                    Severity = TarsEngine.Models.IssueSeverity.Major,
                    Description = "Store credentials in a secure configuration system or environment variables",
                    Location = new CodeLocation
                    {
                        StartLine = lineNumber,
                        EndLine = lineNumber
                    },
                    Title = "Hardcoded Credentials"
                });
            }

            // Detect connection strings with credentials
            var connectionStringRegex = new Regex(@"connectionString\s*=\s*[""'].*?(password|pwd)=[^;]+[""']", RegexOptions.IgnoreCase | RegexOptions.Compiled);
            var connectionStringMatches = connectionStringRegex.Matches(content);
            foreach (Match match in connectionStringMatches)
            {
                var lineNumber = GetLineNumber(content, match.Index);
                issues.Add(new CodeIssue
                {
                    Type = CodeIssueType.Security,
                    Severity = TarsEngine.Models.IssueSeverity.Major,
                    Description = "Store connection strings in a secure configuration system or use a connection string builder with secure parameters",
                    Location = new CodeLocation
                    {
                        StartLine = lineNumber,
                        EndLine = lineNumber
                    },
                    Title = "Connection String with Credentials"
                });
            }
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error detecting hardcoded credentials");
        }

        return issues;
    }

    /// <inheritdoc/>
    public List<CodeIssue> DetectInsecureCryptography(string content)
    {
        var issues = new List<CodeIssue>();

        try
        {
            if (string.IsNullOrWhiteSpace(content))
            {
                return issues;
            }

            // Detect weak hash algorithms
            var weakHashRegex = new Regex(@"(MD5|SHA1)\.Create\(\)", RegexOptions.Compiled);
            var weakHashMatches = weakHashRegex.Matches(content);
            foreach (Match match in weakHashMatches)
            {
                var lineNumber = GetLineNumber(content, match.Index);
                issues.Add(new CodeIssue
                {
                    Type = CodeIssueType.Security,
                    Severity = TarsEngine.Models.IssueSeverity.Major,
                    Description = "Use a stronger hash algorithm like SHA256 or SHA512",
                    Location = new CodeLocation
                    {
                        StartLine = lineNumber,
                        EndLine = lineNumber
                    },
                    Title = "Weak Hash Algorithm"
                });
            }

            // Detect weak encryption algorithms
            var weakEncryptionRegex = new Regex(@"(DES|TripleDES|RC2)\.Create\(\)", RegexOptions.Compiled);
            var weakEncryptionMatches = weakEncryptionRegex.Matches(content);
            foreach (Match match in weakEncryptionMatches)
            {
                var lineNumber = GetLineNumber(content, match.Index);
                issues.Add(new CodeIssue
                {
                    Type = CodeIssueType.Security,
                    Severity = TarsEngine.Models.IssueSeverity.Major,
                    Description = "Use a stronger encryption algorithm like AES",
                    Location = new CodeLocation
                    {
                        StartLine = lineNumber,
                        EndLine = lineNumber
                    },
                    Title = "Weak Encryption Algorithm"
                });
            }

            // Detect insecure random number generation
            var insecureRandomRegex = new Regex(@"new\s+Random\s*\(\s*\)", RegexOptions.Compiled);
            var insecureRandomMatches = insecureRandomRegex.Matches(content);
            foreach (Match match in insecureRandomMatches)
            {
                var lineNumber = GetLineNumber(content, match.Index);
                issues.Add(new CodeIssue
                {
                    Type = CodeIssueType.Security,
                    Severity = TarsEngine.Models.IssueSeverity.Minor,
                    Description = "Use RNGCryptoServiceProvider or RandomNumberGenerator for cryptographic operations",
                    Location = new CodeLocation
                    {
                        StartLine = lineNumber,
                        EndLine = lineNumber
                    },
                    Title = "Insecure Random Number Generation"
                });
            }
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error detecting insecure cryptography");
        }

        return issues;
    }

    /// <inheritdoc/>
    public Dictionary<Interfaces.IssueSeverity, string> GetAvailableSeverities()
    {
        return new Dictionary<Interfaces.IssueSeverity, string>
        {
            { Interfaces.IssueSeverity.Critical, "Critical security vulnerability that must be fixed immediately" },
            { Interfaces.IssueSeverity.Major, "High-risk security vulnerability that should be fixed soon" },
            { Interfaces.IssueSeverity.Minor, "Medium-risk security vulnerability" },
            { Interfaces.IssueSeverity.Trivial, "Low-risk security vulnerability" },
            { Interfaces.IssueSeverity.Warning, "Informational security issue" }
        };
    }

    /// <inheritdoc/>
    public int GetLineNumber(string content, int position)
    {
        if (string.IsNullOrEmpty(content) || position < 0 || position >= content.Length)
        {
            return 0;
        }

        // Count newlines before the position
        return content[..position].Count(c => c == '\n') + 1;
    }
}